File: zig/src/lut.zig
const std = @import("std");

const print = std.debug.print;

fn generateLUT(comptime steps: usize, comptime increment: f64) [steps]f64 {

    // table empty, but enough to hold all the sceps
    var table: [steps]f64 = undefined;

    for (&table, 0..) |*item, i| {
        const result: f64 = @as(f64, @floatFromInt(i)) * increment;
        item.* = @sin(result);
    }

    return table;
}

pub fn main() void {
    const degrees: comptime_float = 360;

    // 1.8MB
    //const increment: comptime_float = 0.1;

    //
    const increment: comptime_float = 0.001;

    const step: comptime_int = @intFromFloat(degrees / increment);

    print("step: {d}\n", .{step});

    const myLut = generateLUT(step, increment);

    const toLookUp = [_]f64{ 15.0, 45.0, 120.0 };

    for (toLookUp) |num| {
        const index: usize = @intFromFloat(num / increment);
        print("{d}: {d}\n", .{ num, myLut[index] });
    }
}

File: rust/src/lut.rs
use trig_const::{cos, sin};

// use rand::prelude::*;

const INCREMENT: f64 = 0.01;
const TEST_SIZE: usize = 500;
const DEGREES: f64 = 360.0;
const STEPS: usize = (DEGREES / INCREMENT) as usize;

const fn generate_lut<const STEPS: usize>() -> [f64; STEPS] {
    // init with all 0's
    let mut arr = [0.0; STEPS];

    let mut i = 0;

    // no compile-time for loops, you have to use a while loop
    while i < STEPS {
        let x = cos(i as f64) + sin(i as f64);
        arr[i] = x;
        i += 1;
    }

    arr
}

fn main() {
    static MY_LUT: [f64; STEPS] = generate_lut();

    for (i, item) in MY_LUT.iter().enumerate() {
        println!("index: {}, value: {}", i, item);
    }
}
